zerver/actions/create_user.py

```diff
@@ -185,18 +180,42 @@ def add_new_user_history(user_profile: UserProfile, streams: Iterable[Stream]) -
     ]
 
     # Start by finding recent messages matching those recipients.
-    cutoff_date = timezone_now() - ONBOARDING_RECENT_TIMEDELTA
+    cutoff_date = timezone_now() - RECENT_MESSAGES_TIMEDELTA
     recent_message_ids = set(
         Message.objects.filter(
             # Uses index: zerver_message_realm_recipient_id
-            realm_id=user_profile.realm_id,
+            realm_id=realm.id,
             recipient_id__in=recipient_ids,
             date_sent__gt=cutoff_date,
         )
         .order_by("-id")
-        .values_list("id", flat=True)[0:MAX_NUM_ONBOARDING_MESSAGES]
+        .values_list("id", flat=True)[0:MAX_NUM_RECENT_MESSAGES]
+    )
+
+    from zerver.lib.onboarding import ONBOARDING_MESSAGES_COUNT
+
+    # IDs of onboarding messages sent by welcome bot.
+    welcome_bot = get_system_bot(settings.WELCOME_BOT, realm.id)
+    onboarding_message_ids = set(
+        # Uses index: zerver_message_realm_sender_recipient
+        Message.objects.filter(
```

Comment:
Hm, it actually uses `zerver_message_realm_recipient_date_sent`:
```sql
explain analyze SELECT "zerver_message"."id" FROM "zerver_message" WHERE ("zerver_message"."realm_id" = 
2 AND "zerver_message"."recipient_id" IN (193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 
207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217) AND "zerver_message"."sender_id" = 5) ORDER BY "zerver_message"."id" ASC LIMIT 12;
```
```sql
 Limit  (cost=15.90..15.90 rows=1 width=4) (actual time=0.040..0.040 rows=0 loops=1)
   ->  Sort  (cost=15.90..15.90 rows=1 width=4) (actual time=0.039..0.039 rows=0 loops=1)
         Sort Key: id
         Sort Method: quicksort  Memory: 25kB
         ->  Index Scan using zerver_message_realm_recipient_date_sent on zerver_message  (cost=0.14..15.89 rows=1 width=4) (actual time=0.033..0.033 rows=0 loops=1)
               Index Cond: ((realm_id = 2) AND (recipient_id = ANY ('{193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217}'::integer[])))
               Filter: (sender_id = 5)
 Planning Time: 0.648 ms
 Execution Time: 0.060 ms
(9 rows)
```

Reply:
Updated the comment.

Reply:
It might be a different index on a production system with more data.

Reply:
A new index `zerver_message_realm_type` has been created for the new algorithm, so marking it as resolved.